events/service/src/main/java/org/projectnessie/events/service/EventSubscribers.java

```diff
@@ -0,0 +1,154 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import org.projectnessie.events.api.EventType;
+import org.projectnessie.events.spi.EventSubscriber;
+import org.projectnessie.events.spi.EventSubscription;
+import org.projectnessie.versioned.ResultType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Loads and holds all {@link EventSubscriber}s.
+ *
+ * <p>The main purpose of this class is to provide a single point of access to all subscribers, and
+ * also to provide a bit-mask of all event types that subscribers are subscribed to, for a fast and
+ * efficient type-based event filtering.
+ *
+ * <p>This class is meant to be used as a singleton, or in CDI Dependent pseudo-scope.
+ */
+public class EventSubscribers implements AutoCloseable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(EventSubscribers.class);
+
+  /** Load all {@link EventSubscriber}s via {@link ServiceLoader}. */
+  public static List<EventSubscriber> loadSubscribers() {
+    List<EventSubscriber> subscribers = new ArrayList<>();
+    for (EventSubscriber subscriber : ServiceLoader.load(EventSubscriber.class)) {
+      subscribers.add(subscriber);
+    }
+    return subscribers;
+  }
+
+  // Visible for testing
+  final List<EventSubscriber> subscribers;
+
+  private final EnumSet<EventType> acceptedEventTypes;
+  private final EnumSet<ResultType> acceptedResultTypes;
+
+  // guarded by this
+  private boolean started;
+  // guarded by this
+  private boolean closed;
+
+  public EventSubscribers(EventSubscriber... subscribers) {
+    this(Arrays.asList(subscribers));
+  }
+
+  public EventSubscribers(Iterable<EventSubscriber> subscribers) {
+    this.subscribers =
+        Collections.unmodifiableList(
+            StreamSupport.stream(subscribers.spliterator(), false).collect(Collectors.toList()));
+    acceptedEventTypes =
+        Arrays.stream(EventType.values())
+            .filter(t -> this.subscribers.stream().anyMatch(s -> s.accepts(t)))
+            .collect(Collectors.toCollection(() -> EnumSet.noneOf(EventType.class)));
+    acceptedResultTypes =
+        acceptedEventTypes.stream()
+            .flatMap(EventSubscribers::map)
+            .collect(Collectors.toCollection(() -> EnumSet.noneOf(ResultType.class)));
+  }
+
+  public void start(Function<EventSubscriber, EventSubscription> subscriptionFactory) {
+    if (!started) {
+      LOGGER.info("Starting subscribers...");
+      for (EventSubscriber subscriber : subscribers) {
+        try {
+          EventSubscription subscription = subscriptionFactory.apply(subscriber);
+          subscriber.onSubscribe(subscription);
+        } catch (Exception e) {
+          LOGGER.error("Error starting subscriber", e);
```

Comment:
```suggestion
          throw new RuntimeException("Error starting subscriber", e);
```
?

Reply:
Not sure... this means that the next subscriber won't be started either.

Reply:
Well, I guess it's safer to prevent the application from starting, rather than having a half-initialized list of subscribers. So OK to do as you suggest.